drivers/sensor/st/lsm9ds1/lsm9ds1.c

```diff
@@ -0,0 +1,677 @@
+
+	ret = lsm9ds1_write_reg(ctx, LSM9DS1_CTRL_REG8, (uint8_t *)&ctrl8_reg, 1);
+	if (ret < 0) {
+		return ret;
+	}
+
+	k_busy_wait(USEC_PER_MSEC * 50U);
+
+	return 0;
+}
+
+static int lsm9ds1_accel_range_to_fs_val(int32_t range)
+{
+	switch (range) {
+	case 2:
+		return LSM9DS1_2g;
+	case 4:
+		return LSM9DS1_4g;
+	case 8:
+		return LSM9DS1_8g;
+	case 16:
+		return LSM9DS1_16g;
+	default:
+		return -EINVAL;
+	}
+}
+
+static int lsm9ds1_gyro_range_to_fs_val(int32_t range)
+{
+	switch (range) {
+	case 245:
+		return LSM9DS1_245dps;
+	case 500:
+		return LSM9DS1_500dps;
+	case 2000:
+		return LSM9DS1_2000dps;
+	default:
+		return -EINVAL;
+	}
+}
+
+static int lsm9ds1_accel_fs_val_to_gain(int fs)
+{
+
+	return lsm9ds1_accel_fs_sens[fs];
+}
+
+static int lsm9ds1_accel_freq_to_odr_val(uint16_t freq)
+{
+	size_t i;
+
+	for (i = 0; i < ARRAY_SIZE(lsm9ds1_odr_map); i++) {
+		if (freq <= lsm9ds1_odr_map[i]) {
+			return i;
+		}
+	}
+
+	return -EINVAL;
+}
+
+static int lsm9ds1_gyro_freq_to_odr_val(uint16_t freq)
+{
+	size_t i;
+
+	for (i = 0; i < ARRAY_SIZE(lsm9ds1_gyro_odr_map); i++) {
+		if (freq <= lsm9ds1_gyro_odr_map[i]) {
+			return i;
+		}
+	}
+
+	return -EINVAL;
+}
+
+static int lsm9ds1_accel_set_odr_raw(const struct device *dev, uint8_t odr)
+{
+	const struct lsm9ds1_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	struct lsm9ds1_data *data = dev->data;
+	int ret;
+
+	lsm9ds1_ctrl_reg6_xl_t ctrl_reg6_xl;
+
+	ret = lsm9ds1_read_reg(ctx, LSM9DS1_CTRL_REG6_XL,
+				(uint8_t *)&ctrl_reg6_xl, 1);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ctrl_reg6_xl.odr_xl = odr;
+
+	ret = lsm9ds1_write_reg(ctx, LSM9DS1_CTRL_REG6_XL,
+				(uint8_t *)&ctrl_reg6_xl, 1);
+	if (ret < 0) {
+		return ret;
+	}
+
+	data->accel_odr = odr;
+
+	return 0;
+}
+
+static int lsm9ds1_gyro_set_odr_raw(const struct device *dev, uint8_t odr)
+{
+	const struct lsm9ds1_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	struct lsm9ds1_data *data = dev->data;
+	int ret;
+
+	lsm9ds1_ctrl_reg1_g_t ctrl_reg1;
+
+	ret = lsm9ds1_read_reg(ctx, LSM9DS1_CTRL_REG1_G,
+				(uint8_t *)&ctrl_reg1, 1);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ctrl_reg1.odr_g = odr;
+
+	ret = lsm9ds1_write_reg(ctx, LSM9DS1_CTRL_REG1_G,
+				(uint8_t *)&ctrl_reg1, 1);
+	if (ret < 0) {
+		return ret;
+	}
+
+	data->gyro_odr = odr;
+	return 0;
+
+}
+
+static int lsm9ds1_gyro_odr_set(const struct device *dev, uint16_t freq)
+{
+	struct lsm9ds1_data *data = dev->data;
+	int odr;
+	int ret;
+
+	odr = lsm9ds1_gyro_freq_to_odr_val(freq);
+
+	if (odr == data->gyro_odr) {
+		return 0;
+	}
+
+	LOG_INF("You are also changing the odr of the accelerometer");
+
+	ret = lsm9ds1_gyro_set_odr_raw(dev, odr);
+	if (ret < 0) {
+		LOG_DBG("failed to set gyroscope sampling rate");
+		return ret;
+	}
+
+	/*
+	 * When the gyroscope is on, the value of the accelerometer odr must be
+	 * the same as the value of the gyroscope.
+	 */
+
+	ret = lsm9ds1_accel_set_odr_raw(dev, odr);
```

Comment:
What happens if `15` is chosen here when accel seems to want `10`? Would it make sense to check for equal values before changing?

Reply:
When the gyroscope is on, the frequencies of the gyroscope are the one that matters.
So if the gyroscope is off, accel can be at 10 Hz, if the gyroscope is on, accel will be at 15 Hz (or another frequency on the gyroscope list), and can't be at 10 Hz as long as the gyroscope is on.
In fact, if the gyroscope is on, the value of the odr register of the accel is ignored, but the documentation and the HAL wants to have the same value in both registers, so I set it to the right value.